rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingAlg.scala

```diff
@@ -1,44 +1,39 @@
 package coop.rchain.rholang.interpreter.accounting
 
-import cats.Monad
 import cats.effect.Sync
 import cats.effect.concurrent.Ref
-import coop.rchain.rholang.interpreter.accounting
 import cats.implicits._
+import cats.{FlatMap, Monad}
+import coop.rchain.rholang.interpreter.errors.OutOfPhlogistonsError
 
 trait CostAccountingAlg[F[_]] {
   def charge(cost: Cost): F[Unit]
-  def modify(f: CostAccount => CostAccount): F[Unit]
-  def getCost(): F[CostAccount]
-  def setCost(cost: CostAccount): F[Unit]
+  def get(): F[CostAccount]
+  def set(cost: CostAccount): F[Unit]
```

Comment:
This really should just be `def charge(cost: CostAccount) : F[Unit]`.
The only use case for set is initialization before running the computation.
The only use case for get is getting the result past the computation.
All other occurrences of get & set are equivalent to ~`.charge(costAlg.get)`

To satisfy the init + get result use cases, we should make so that the CostAccountingAlg instances do not hide the `Ref`-s internally, but use a ref that's provided by the code creating the instance. Notice this is not leaking internals, as the ref's state is fully accessible via the .get .set right now. Quite the contrary - instantiating based on a passed ref will allow **limiting** the places where the Ref's state can be accessed to just the end of the world and the implementation of `charge`.

Reply:
That's a part of what I meant, and the reason why we can make the types guard that :)

Reply:
I don't understand why `Ref` would be better.

Reply:
Because, currently in every part of spatial matcher or other place whrere I have CostAccountAlg, I can break the cost account by using set.

When we only have `CostAccountAlg.charge(Cost|CostAccount)` in the interface, this is no longer possible.

This is Interface Segragation Principle if you will: we separate the initialization capabilities (which can break the accounting!) from charging capabilities (which are the only that are needed where CostAccountAlg is used). The place where it's created and the costs are needed to be read is **another** usage site, with different use cases, which demands a different API.

Pardon the OOPsplanation ;)

Reply:
Makes sense, I will see what I can do.

Reply:
I've noticed there will be some churn around marrying CostCaccountAlg and SpatialMatcher (which both do cost accounting...). But not that much:
We'll simply expose a '.get' method on CostAccountAlg and will use 'charge' instead of 'set'. Which happens to be naturally supported by accumulating cost towards limit (vs subtracting from a limit) - we simply get the 'cost so far' from the matcher and do costAccountingAlg.charge(matcherCost)